File: src/main/java/phoenix/utils/RenderUtils.java
package phoenix.utils;

import com.mojang.blaze3d.systems.RenderSystem;
import com.mojang.blaze3d.vertex.IVertexBuilder;
import javafx.util.Pair;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.AbstractGui;
import net.minecraft.client.renderer.BufferBuilder;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.WorldVertexBufferUploader;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import net.minecraft.util.ResourceLocation;

public class RenderUtils
{
    public static void drawWithScale(ResourceLocation texture, int sizeX, int sizeY, int x, int y, float scale)
    {
        RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
        Minecraft.getInstance().getTextureManager().bindTexture(texture);
        RenderSystem.scaled(scale, scale, scale);//увеличиваем картинку
        blit(x, y, 0, 0, sizeX, sizeY);
        RenderSystem.scaled(1 / scale, 1 / scale, 1 / scale);//возвращаем старый скейл, чтоб тект был нормальным
    }

    public static void drawRectScalable
            (ResourceLocation texture,
             int x, int y,
             double maxSizeX, double maxSizeY,
             int depth)
    {
        Minecraft.getInstance().getTextureManager().bindTexture(texture);
        Pair<Integer, Integer> d = TextureUtils.getTextureSize(texture);
        RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
        double scale = maxSizeX / d.getKey();
        int sizeX = (int) (d.getKey() * scale),
            sizeY = (int) (d.getValue() * scale);
        if(maxSizeY < sizeY)
        {
            scale *= maxSizeY / sizeY;
            sizeY = (int) (d.getKey() * scale);
            sizeX = (int) (d.getValue() * scale);
        }

        blit(x, y, depth, 0, 0, sizeX, sizeY, sizeX, sizeY);
    }

    public static void drawTexturedRectWithSize(ResourceLocation texture, int textureSizeX, int textureSizeY, int x, int y, int sizeX, int sizeY, int blitOffset)
    {
        RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
        Minecraft.getInstance().getTextureManager().bindTexture(texture);
        blit(x, y, blitOffset, 0, 0, sizeX, sizeY, textureSizeX, textureSizeY);
    }

    public static void drawTexturedRectWithSize(ResourceLocation texture, int textureSizeX, int textureSizeY, int x, int y, int sizeX, int sizeY, int blitOffset, int u, int v)
    {
        RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
        Minecraft.getInstance().getTextureManager().bindTexture(texture);
        AbstractGui.blit(x, y, blitOffset, (float) u, (float) v, sizeX, sizeY, textureSizeX, textureSizeY);
    }

    public static void blit(int x, int y, float offsetU, float offsetV, int sizeX, int sizeY)
    {
        blit(x, y, 0, offsetU, offsetV, sizeX, sizeY, 256, 256);
    }

    public static void blit(int x, int y, int depth, float offsetU, float offsetV, int sizeX, int sizeY, int texSizeX, int texSizeY)
    {
        draw(x, x + sizeX, y, y + sizeY, depth, offsetU, offsetV, texSizeX, texSizeY, sizeX, sizeY);
    }

    private static void draw(int posA, int posB, int posC, int posD, int depth, float offsetU, float offsetV, float texSizeX, float texSizeY, int sizeX, int sizeY)
    {
        //*
        innerBlit(posA, posB, posC, posD, depth,
                (offsetU + 0.00F) / texSizeX,
                (offsetU + sizeX) / texSizeX,
                (offsetV + 0.00F) / texSizeY,
                (offsetV + sizeY) / texSizeY);

        //*/
        /*
        innerBlit(posA, posB, posC, posD, depth,
                (offsetU + 0.00F) / texSizeX,
                (offsetU + texSizeX) / texSizeX,
                (offsetV + 0.00F) / texSizeY,
                (offsetV + texSizeY) / texSizeY);
        //*/
    }

    public static void innerBlit(int posA, int posB, int posC, int posD, int depth, float uvA, float uvB, float uvC, float uvD)
    {
        BufferBuilder bufferbuilder = Tessellator.getInstance().getBuffer();
        bufferbuilder.begin(7, DefaultVertexFormats.POSITION_TEX);
        bufferbuilder.pos(posA, posD, depth).tex(uvA, uvD).endVertex();
        bufferbuilder.pos(posB, posD, depth).tex(uvB, uvD).endVertex();
        bufferbuilder.pos(posB, posC, depth).tex(uvB, uvC).endVertex();
        bufferbuilder.pos(posA, posC, depth).tex(uvA, uvC).endVertex();
        bufferbuilder.finishDrawing();
        RenderSystem.enableAlphaTest();
        WorldVertexBufferUploader.draw(bufferbuilder);
    }

    public static void refreshDrawing(IVertexBuilder vb, RenderType type) {
        if (vb instanceof BufferBuilder) {
            type.finish((BufferBuilder) vb, 0, 0, 0);
            ((BufferBuilder) vb).begin(type.getDrawMode(), type.getVertexFormat());
        }
    }
}

File: src/main/java/phoenix/utils/RenderUtils.kt
package phoenix.utils

import com.mojang.blaze3d.systems.RenderSystem
import com.mojang.blaze3d.vertex.IVertexBuilder
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.AbstractGui
import net.minecraft.client.renderer.BufferBuilder
import net.minecraft.client.renderer.RenderType
import net.minecraft.client.renderer.Tessellator
import net.minecraft.client.renderer.WorldVertexBufferUploader
import net.minecraft.client.renderer.vertex.DefaultVertexFormats
import net.minecraft.util.ResourceLocation
import phoenix.utils.TextureUtils.getTextureSize


object RenderUtils
{
    fun drawWithScale(texture: ResourceLocation, sizeX: Int, sizeY: Int, x: Int, y: Int, scale: Float)
    {
        RenderSystem.color4f(1.0f, 1.0f, 1.0f, 1.0f)
        Minecraft.getInstance().getTextureManager().bindTexture(texture)
        RenderSystem.scaled(scale.toDouble(), scale.toDouble(), scale.toDouble()) //увеличиваем картинку
        blit(x, y, 0f, 0f, sizeX, sizeY)
        RenderSystem.scaled(
            (1 / scale).toDouble(),
            (1 / scale).toDouble(),
            (1 / scale).toDouble()
        ) //возвращаем старый скейл, чтоб тект был нормальным
    }

    fun drawRectScalable(
        texture: ResourceLocation,
        x: Int, y: Int,
        maxSizeX: Double, maxSizeY: Double,
        depth: Int
    )
    {
        Minecraft.getInstance().getTextureManager().bindTexture(texture)
        val d = getTextureSize(texture)
        RenderSystem.color4f(1.0f, 1.0f, 1.0f, 1.0f)
        var scale = maxSizeX / d.key
        var sizeX = (d.key * scale).toInt()
        var sizeY = (d.value * scale).toInt()
        if (maxSizeY < sizeY)
        {
            scale *= maxSizeY / sizeY
            sizeY = (d.key * scale).toInt()
            sizeX = (d.value * scale).toInt()
        }
        blit(x, y, depth, 0f, 0f, sizeX, sizeY, sizeX, sizeY)
    }

    fun drawTexturedRectWithSize(
        texture: ResourceLocation,
        textureSizeX: Int,
        textureSizeY: Int,
        x: Int,
        y: Int,
        sizeX: Int,
        sizeY: Int,
        blitOffset: Int
    )
    {
        RenderSystem.color4f(1.0f, 1.0f, 1.0f, 1.0f)
        Minecraft.getInstance().getTextureManager().bindTexture(texture)
        blit(x, y, blitOffset, 0f, 0f, sizeX, sizeY, textureSizeX, textureSizeY)
    }

    fun drawTexturedRectWithSize(
        texture: ResourceLocation,
        textureSizeX: Int,
        textureSizeY: Int,
        x: Int,
        y: Int,
        sizeX: Int,
        sizeY: Int,
        blitOffset: Int,
        u: Int,
        v: Int
    )
    {
        RenderSystem.color4f(1.0f, 1.0f, 1.0f, 1.0f)
        Minecraft.getInstance().getTextureManager().bindTexture(texture)
        AbstractGui.blit(x, y, blitOffset, u.toFloat(), v.toFloat(), sizeX, sizeY, textureSizeX, textureSizeY)
    }

    fun blit(x: Int, y: Int, offsetU: Float, offsetV: Float, sizeX: Int, sizeY: Int)
    {
        blit(x, y, 0, offsetU, offsetV, sizeX, sizeY, 256, 256)
    }

    fun blit(
        x: Int,
        y: Int,
        depth: Int,
        offsetU: Float,
        offsetV: Float,
        sizeX: Int,
        sizeY: Int,
        texSizeX: Int,
        texSizeY: Int
    )
    {
        draw(x, x + sizeX, y, y + sizeY, depth, offsetU, offsetV, texSizeX.toFloat(), texSizeY.toFloat(), sizeX, sizeY)
    }

    private fun draw(
        posA: Int,
        posB: Int,
        posC: Int,
        posD: Int,
        depth: Int,
        offsetU: Float,
        offsetV: Float,
        texSizeX: Float,
        texSizeY: Float,
        sizeX: Int,
        sizeY: Int
    )
    {
        //*
        innerBlit(
            posA, posB, posC, posD, depth,
            (offsetU + 0.00f) / texSizeX,
            (offsetU + sizeX) / texSizeX,
            (offsetV + 0.00f) / texSizeY,
            (offsetV + sizeY) / texSizeY
        )

        //*/
        /*
        innerBlit(posA, posB, posC, posD, depth,
                (offsetU + 0.00F) / texSizeX,
                (offsetU + texSizeX) / texSizeX,
                (offsetV + 0.00F) / texSizeY,
                (offsetV + texSizeY) / texSizeY);
        // */
    }

    fun innerBlit(
        posA: Int,
        posB: Int,
        posC: Int,
        posD: Int,
        depth: Int,
        uvA: Float,
        uvB: Float,
        uvC: Float,
        uvD: Float
    )
    {
        val bufferbuilder = Tessellator.getInstance().buffer
        bufferbuilder.begin(7, DefaultVertexFormats.POSITION_TEX)
        bufferbuilder.pos(posA.toDouble(), posD.toDouble(), depth.toDouble()).tex(uvA, uvD).endVertex()
        bufferbuilder.pos(posB.toDouble(), posD.toDouble(), depth.toDouble()).tex(uvB, uvD).endVertex()
        bufferbuilder.pos(posB.toDouble(), posC.toDouble(), depth.toDouble()).tex(uvB, uvC).endVertex()
        bufferbuilder.pos(posA.toDouble(), posC.toDouble(), depth.toDouble()).tex(uvA, uvC).endVertex()
        bufferbuilder.finishDrawing()
        RenderSystem.enableAlphaTest()
        WorldVertexBufferUploader.draw(bufferbuilder)
    }

    fun refreshDrawing(vb: IVertexBuilder, type: RenderType)
    {
        if (vb is BufferBuilder)
        {
            type.finish(vb, 0, 0, 0)
            vb.begin(type.getDrawMode(), type.getVertexFormat())
        }
    }
}
